contracts/Staking.sol

```diff
@@ -458,66 +446,110 @@ contract Staking is Autopetrified, ERCStaking, ERCStakingHistory, IStakingLockin
         totalStakedHistory.add64(getBlockNumber64(), newStake);
     }
 
-    /**
-     * Note: So far this function is called from unlock and from transferFromLock,
-     * which both ensure that lock can be unlocked (the latter through isLockManager modifier)
-     */
-    function _unlock(address _accountAddress, uint256 _lockId) internal {
+    function _unsafeUnlock(address _accountAddress, address _lockManager) internal {
         Account storage account = accounts[_accountAddress];
-        Lock storage lock_ = account.locks[_lockId];
+        Lock storage lock_ = account.locks[_lockManager];
 
-        lock_.unlockedAt = getTimestamp64();
+        // update total
+        // no need for SafeMath: totalLocked must be greater or equal than lock._amount
+        account.totalLocked = account.totalLocked - lock_.amount;
 
-        // remove from active locks, replacing it by the last one in the array
-        // we assume consistency here, i.e., that lock exists in active array
-        uint256 locksLength = account.activeLockIds.length;
-        if (locksLength == 1) {
-            delete account.activeLockIds;
-            return;
-        }
+        emit Unlocked(_accountAddress, _lockManager, lock_.amount);
 
-        for (uint256 i = 0; i < locksLength; i++) {
-            if (account.activeLockIds[i] == _lockId) {
-                account.activeLockIds[i] = account.activeLockIds[locksLength - 1];
-                delete account.activeLockIds[locksLength - 1];
-                account.activeLockIds.length--;
-                break;
-            }
-        }
+        delete account.locks[_lockManager];
```

Comment:
Note that this is also removing the allowance, not sure if that's correct

Reply:
Yes, that was the intention, at least when it comes from `unlockAndRemoveManager`. As you pointed out in another comment, it probably doesn’t make sense when coming from `_decreaseActiveLockAmount`.